File: RxTango/python/examples/calibration_pipeline.py
"""Read → calibrate → write pipeline (one shot).

Reads `double_scalar`, applies a linear calibration, writes the result to
`double_scalar_w`, formats it as a string, and writes to `string_scalar_w`.

Mirrors `CalibrationPipeline.java`.

Usage:
    python calibration_pipeline.py [device]

    device  defaults to tango://localhost:10000/sys/tg_test/1
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import read_attribute, write_attribute


async def main() -> None:
    device = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)
    done      = asyncio.Event()

    print(f"Calibration pipeline on {device}\n")

    read_attribute(device, "double_scalar").pipe(
        ops.do_action(on_next=lambda v: print(f"  [1] read    double_scalar  = {v:+.6f}")),

        # Linear calibration: y = |x| * 2.0 + 1.5
        ops.map(lambda v: abs(v) * 2.0 + 1.5),
        ops.do_action(on_next=lambda v: print(f"  [2] calibrated             = {v:+.6f}")),

        # Write calibrated value
        ops.flat_map(lambda v: write_attribute(device, "double_scalar_w", v)),
        ops.do_action(on_next=lambda v: print(f"  [3] wrote   double_scalar_w = {v:+.6f}")),

        # Format as string
        ops.map(lambda v: f"cal={v:.4f}"),
        ops.do_action(on_next=lambda s: print(f"  [4] formatted              = {s!r}")),

        # Write string (write_attribute works for any type PyTango accepts)
        ops.flat_map(lambda s: write_attribute(device, "string_scalar_w", s)),
        ops.do_action(on_next=lambda s: print(f"  [5] wrote   string_scalar_w = {s!r}")),

    ).subscribe(
        on_next=lambda _: None,
        on_error=lambda e: (print(f"  ERROR: {e}", file=sys.stderr), done.set()),
        on_completed=done.set,
        scheduler=scheduler,
    )

    await asyncio.wait_for(done.wait(), timeout=10.0)
    print("\n  Pipeline complete.")


if __name__ == "__main__":
    asyncio.run(main())

File: RxTango/cpp/examples/calibration_pipeline.cpp
/**
 * Continuous calibration pipeline — read → calibrate → write, forever.
 *
 * The same three reactive operators Java/Python use — interval · flat_map · map:
 *   interval  →  read double_scalar
 *              →  calibrate (abs(v) * 2.0 + 1.5)
 *              →  write double_scalar_w
 *
 * Mirrors Python's calibration_pipeline.py and Java's CalibrationPipeline.java.
 *
 * Usage:
 *   ./calibration_pipeline [device] [interval-ms]
 *   defaults: sys/tg_test/1  500
 */

#include <chrono>
#include <csignal>
#include <cmath>
#include <iostream>
#include <string>

#include <rxcpp/rx.hpp>
#include <rxtango/rxtango.hpp>

static volatile sig_atomic_t g_running = 1;

int main(int argc, char* argv[]) {
    const std::string device      = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";
    const int         interval_ms = argc > 2 ? std::stoi(argv[2]) : 500;

    std::signal(SIGINT, [](int) { g_running = 0; });

    std::cout << "Calibration pipeline on " << device
              << "  every " << interval_ms << " ms  (Ctrl+C to stop)\n\n";

    auto sub = rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
        .flat_map([device](long) {
            return rxtango::read_attribute<double>(device, "double_scalar");
        })
        .map([](double v) { return std::abs(v) * 2.0 + 1.5; })
        .flat_map([device](double calibrated) {
            return rxtango::write_attribute<double>(device, "double_scalar_w", calibrated);
        })
        .subscribe(
            [](double written) {
                std::cout << "  wrote: " << written << "\n";
            },
            [](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
            }
        );

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    sub.unsubscribe();
    return 0;
}
